luna.c

```diff
@@ -524,52 +524,90 @@ int add_default_document_to_tns(const char *tnsfile_path, unsigned tiversion) {
-			goto add_infile_err;
-		memcpy(header_and_deflated_buf, tien_crypted_header, header_size);
-		if (add_processed_file_to_tns(filename, header_and_deflated_buf, header_size + deflated_size, tnsfile_path, tiversion))
-			goto add_infile_err;
-	}
-	else { // don't crypt, don't deflate: will be deflated by minizip
-		if (add_processed_file_to_tns(filename, xmlc_buf, xmlc_buf_size, tnsfile_path, tiversion))
-			goto add_infile_err;
-	}
+
+	uint8_t *def_buf = header_and_deflated_buf + header_size;
+	long deflated_size = deflate_compressed_xml(def_buf, def_size, xmlc_buf, xmlc_buf_size);
+	if (doccrypt(def_buf, deflated_size))
+		goto add_compressed_xml_err;
+	memcpy(header_and_deflated_buf, tien_crypted_header, header_size);
+	if (add_processed_file_to_tns(filename, header_and_deflated_buf, header_size + deflated_size, tnsfile_path, tiversion))
+		goto add_compressed_xml_err;
 	free(header_and_deflated_buf);
 	return 0;
 }
 
+int add_infile_to_tns(const char *infile_path, const char *tnsfile_path, unsigned tiversion) {
+	size_t xmlc_buf_size;
+	const char *filename = NULL;
+	void *xmlc_buf = read_file_and_xml_compress(infile_path, &xmlc_buf_size, &filename);
+	int ret;
+	if (!xmlc_buf)
+		return 1;
+
+	if (has_ext(filename, ".xml")) // Only .xml files are encrypted
+		ret = add_compressed_xml_to_tns(tnsfile_path, filename, xmlc_buf, xmlc_buf_size, tiversion);
+	else // don't crypt, don't deflate: will be deflated by minizip
+		ret = add_processed_file_to_tns(filename, xmlc_buf, xmlc_buf_size, tnsfile_path, tiversion);
+	free(xmlc_buf);
+	return ret;
+}
+
+// Add XML that refers to the Python script. The Python script itself is added separately.
+int add_python_xml_to_tns(const char *python_path, const char *tnsfile_path, unsigned tiversion) {
+	static const char py_header[] =
+		"TIXC0100-1.0?><prob xmlns=\"urn:TI.Problem\" ver=\"1.0\" pbname=\"\">"
+		"<sym>\x0E\x01<card clay=\"0\" h1=\"10000\" h2=\"10000\" w1=\"10000\" "
+		"w2=\"10000\"><isDummyCard>0\x0E\x03<flag>0\x0E\x04<wdgt xmlns:py=\"urn:"
+		"TI.PythonEditor\" type=\"TI.PythonEditor\" ver=\"1.0\"><py:data><py:name>";
+	static const char py_footer[] =
+		"\x0E\x07<py:dirf>-10000000\x0E\x08\x0E\x06<py:mFlags>1024\x0E\x09"
+		"<py:value>10\x0E\x0A\x0E\x05\x0E\x02\x0E\x00";
+
+	const char *filename = gnu_basename(python_path);
+	size_t filename_len = strlen(filename);
+	if (filename_len > 240) {
+		puts("Python script filenames limited to 240 characters");
+		return 1;
+	}
+
+	// Filename goes sandwiched between the header and footer
+	size_t py_header_size = sizeof(py_header) - 1;
+	size_t py_footer_size = sizeof(py_footer) - 1;
+	size_t total_size = py_header_size + filename_len + py_footer_size;
+	uint8_t xmlc_buf[(sizeof(py_header) - 1) + 240 + (sizeof(py_footer - 1))];
+
+	memcpy(xmlc_buf, py_header, py_header_size);
+	memcpy(xmlc_buf + py_header_size, filename, filename_len);
+	memcpy(xmlc_buf + py_header_size + filename_len, py_footer, py_footer_size);
```

Comment:
What about using a simple `snprintf`/`asprintf` for all of the above?

Reply:
Good idea. Actually I just changed it to sprintf, since the buffer is guaranteed to be large enough.